analytic_engine/src/instance/serializer.rs

```diff
@@ -0,0 +1,175 @@
+// Copyright 2023 CeresDB Project Authors. Licensed under Apache-2.0.
+
+use std::{
+    sync::{Arc, Condvar, Mutex},
+    time::Instant,
+};
+
+use common_util::{runtime::Runtime, time::InstantExt};
+use futures::Future;
+use log::error;
+use table_engine::table::{Error as TableError, Result as TableResult, TableId};
+use tokio::sync::oneshot;
+
+use crate::{
+    instance::flush_compaction::{self, BackgroundFlushFailed},
+    table::metrics::Metrics,
+};
+
+#[derive(Default)]
+enum FlushState {
+    #[default]
+    Ok,
+    Flushing,
+    Failed {
+        err_msg: String,
+    },
+}
+
+type ScheduleLock = Arc<(Mutex<FlushState>, Condvar)>;
+
+#[derive(Default)]
+pub struct TableFlushScheduler {
+    schedule_lock: ScheduleLock,
+}
+
+/// To ensure the consistency of a table's data, these rules are required:
+/// - The write procedure (write wal + write memtable) should be serialized as a
+///   whole, that is to say, it is not allowed to write wal and memtable
+///   concurrently or interleave the two sub-procedures;
+/// - Any operation that may change the data of a table should be serialized,
+///   including altering table schema, dropping table, etc;
+/// - The flush procedure of a table should be serialized;
+pub struct TableOpSerializer {
+    table_id: TableId,
+    flush_scheduler: TableFlushScheduler,
+}
+
+impl TableOpSerializer {
+    pub fn new(table_id: TableId) -> Self {
+        Self {
+            table_id,
+            flush_scheduler: TableFlushScheduler::default(),
+        }
+    }
+
+    #[inline]
+    pub fn table_id(&self) -> TableId {
+        self.table_id
+    }
+}
+
+impl TableOpSerializer {
+    pub fn flush_scheduler(&mut self) -> &mut TableFlushScheduler {
+        &mut self.flush_scheduler
+    }
+}
+
+impl TableFlushScheduler {
+    /// Control the flush procedure and ensure multiple flush procedures to be
+    /// sequential.
+    ///
+    /// REQUIRE: should only be called by the write thread.
+    #[allow(clippy::too_many_arguments)]
+    pub async fn flush_sequentially<F, T>(
+        &mut self,
```

Comment:
If this method is `&mut self`, then this method is sequential execution, no concurrency will happen, right?

Reply:
Yes.